### ISSUE TITLE
dataframe support for dot, cov and corr

### ISSUE THREAD

**Issue by DSLituiev** [dask/dask#1000]
Running above mentioned matrix operations on `dask.dataframe` of floats induces errors, e.g. `dot`

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/home/dima/data/external/pandas/pandas/index.pyx in pandas.index.IndexEngine.get_loc (pandas/index.c:3869)()
    156         try:
--> 157             return self.mapping.get_item(val)
    158         except TypeError:

/home/dima/data/external/pandas/pandas/hashtable.pyx in pandas.hashtable.Int64HashTable.get_item (pandas/hashtable.c:6503)()
    302 
--> 303     cpdef get_item(self, int64_t val):
    304         cdef khiter_t k

TypeError: an integer is required

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
/usr/local/lib/python3.4/dist-packages/dask/dataframe/core.py in __getattr__(self, key)
   1309         try:
-> 1310             return self[key]
   1311         except KeyError as e:

/usr/local/lib/python3.4/dist-packages/dask/dataframe/core.py in __getitem__(self, key)
   1274             # error is raised from pandas
-> 1275             dummy = self._pd[_extract_pd(key)]
   1276 

/home/dima/data/external/pandas/pandas/core/frame.py in __getitem__(self, key)
   1979         else:
-> 1980             return self._getitem_column(key)
   1981 

/home/dima/data/external/pandas/pandas/core/frame.py in _getitem_column(self, key)
   1986         if self.columns.is_unique:
-> 1987             return self._get_item_cache(key)
   1988 

/home/dima/data/external/pandas/pandas/core/generic.py in _get_item_cache(self, item)
   1090         if res is None:
-> 1091             values = self._data.get(item)
   1092             res = self._box_item_values(item, values)

/home/dima/data/external/pandas/pandas/core/internals.py in get(self, item, fastpath)
   3193             if not isnull(item):
-> 3194                 loc = self.items.get_loc(item)
   3195             else:

/home/dima/data/external/pandas/pandas/indexes/base.py in get_loc(self, key, method, tolerance)
   1863             key = _values_from_object(key)
-> 1864             return self._engine.get_loc(key)
   1865 

/home/dima/data/external/pandas/pandas/index.pyx in pandas.index.IndexEngine.get_loc (pandas/index.c:4005)()
    136 
--> 137     cpdef get_loc(self, object val):
    138         if is_definitely_invalid_key(val):

/home/dima/data/external/pandas/pandas/index.pyx in pandas.index.IndexEngine.get_loc (pandas/index.c:3934)()
    158         except TypeError:
--> 159             raise KeyError(val)
    160 

KeyError: 'dot'

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
<ipython-input-59-2983061be677> in <module>()
----> 1 gtcov = gtdf.dot( gtdf.T)

/usr/local/lib/python3.4/dist-packages/dask/dataframe/core.py in __getattr__(self, key)
   1310             return self[key]
   1311         except KeyError as e:
-> 1312             raise AttributeError(e)
   1313 
   1314     def __dir__(self):

AttributeError: 'dot'

```

**Comment by mrocklin** [comment]
Yes, dask.dataframe does not yet implement `dot`.  From your experience I see two problems

```python
In [2]: dd.DataFrame.dot
AttributeError: type object 'DataFrame' has no attribute 'dot'
```

1.  We need to improve our error messages for missing attributes/functions on DataFrame objects
2.  We should probably implement `dot` for the common case.

**Comment by jcrist** [comment]
I have implemented `corr` and `cov` in in #1037. I'm not sure if there's a good way to implement `dot` though. Since we don't keep track of the index as part of the metadata, the output column names can't be determined lazily.

Also, I'm unable to reproduce the long stack trace for missing attributes:

```python
In [3]: type(df2)
Out[3]: dask.dataframe.core.DataFrame

In [4]: df2.dot
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-526d781ca1d7> in <module>()
----> 1 df2.dot

/Users/jcrist/Code/dask/dask/dataframe/core.pyc in __getattr__(self, key)
   1358             return self[key]
   1359         except KeyError as e:
-> 1360             raise AttributeError(e)
   1361
   1362     def __dir__(self):

AttributeError: 'dot'
```

**Comment by mrocklin** [comment]
So, if you feel like living dangerously you can always do `(df.x * df.y).sum()` to get `df.x.dot(df.y)`.  We're already fairly dangerous.  I'm going to close this now that the difficult functions (`corr`, `cov`) have been implemented.   Feel free to reopen if having an explicit `dot` is still critical.
